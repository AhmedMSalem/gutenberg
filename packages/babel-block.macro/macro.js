/**
 * External dependencies
 */
const { createMacro } = require( 'babel-plugin-macros' );
const { existsSync, readFileSync } = require( 'fs' );
const { mapKeys, pick } = require( 'lodash' );
const { dirname, join, relative } = require( 'path' );

function getFilename( [ filenamePath ], state ) {
	const filename = filenamePath.evaluate().value;

	return join(
		relative( process.cwd(), dirname( state.file.opts.filename ) ),
		filename
	);
}

function readMetadata( filename ) {
	if ( ! existsSync( filename ) ) {
		throw new Error( `Invalid file name provided: ${ filename }.` );
	}

	const metadataRaw = readFileSync( filename, 'utf8' );

	return JSON.parse( metadataRaw );
}

function formatMetadata( metadata, types ) {
	const aliases = {
		styleVariations: 'styles',
	};
	const replaceWithAlias = ( _, key ) => {
		return aliases[ key ] || key;
	};
	const whitelistedProperties = [
		'name',
		'title',
		'category',
		'parent',
		'icon',
		'description',
		'keywords',
		'attributes',
		'styles',
	];

	const metadataFiltered = pick(
		mapKeys( metadata, replaceWithAlias ),
		whitelistedProperties
	);

	const metadataNode = types.valueToNode( metadataFiltered );
	/*const translatedProperties = [
		'title',
		'description',
		'keywords',
	];*/

	return metadataNode;
}

function babelBlockMacro( { references, state, babel } ) {
	const { types } = babel;
	references.default.forEach( ( referencePath ) => {
		if ( referencePath.parentPath.type === 'CallExpression' ) {
			const metadata = readMetadata(
				getFilename( referencePath.parentPath.get( 'arguments' ), state )
			);

			referencePath.parentPath.replaceWith( formatMetadata( metadata, types ) );
		} else {
			throw new Error(
				`@wordpress/babel-block.macro can only be used as function call. You tried ${ referencePath.parentPath.type }.`,
			);
		}
	} );
}

module.exports = createMacro( babelBlockMacro );
