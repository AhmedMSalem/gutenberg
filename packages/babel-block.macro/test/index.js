/**
 * External dependencies
 */
import plugin from 'babel-plugin-macros';
import pluginTester from 'babel-plugin-tester';

pluginTester( {
	plugin,
	babelOptions: {
		babelrc: false,
		filename: __filename,
		presets: [ '@wordpress/babel-preset-default' ],
	},
	tests: {
		'valid metadata file name': {
			code: `
				import macro from '../macro';
				const metadata = macro( './fixtures/block.json' );
			`,
			snapshot: true,
		},
		'invalid metadata file name': {
			code: `
				import macro from '../macro';
				const metadata = macro( './invalid-file.json' );
			`,
			error: 'Invalid file name provided: packages/babel-block.macro/test/invalid-file.json.',
		},
		'invalid usage: as function argument': {
			code: `
				import macro from '../macro';
        		const metadata = doSomething( macro );
      		`,
			error: true,
		},
		'invalid usage: missing file path': {
			code: `
				import macro from '../macro';
				const metadata = macro;
			`,
			error: true,
		},
	},
} );
