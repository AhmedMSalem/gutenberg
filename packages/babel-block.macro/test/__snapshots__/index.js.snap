// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros valid metadata file name: valid metadata file name 1`] = `
"
import macro from '../macro';
const metadata = macro( './fixtures/block.json' );

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

var metadata = {
  name: \\"my-plugin/notice\\",
  title: \\"Notice\\",
  category: \\"common\\",
  parent: [\\"core/group\\"],
  icon: \\"star\\",
  description: \\"Shows warning, error or success notices  ...\\",
  keywords: [\\"alert\\", \\"message\\"],
  attributes: {
    message: {
      type: \\"string\\",
      source: \\"html\\",
      selector: \\".meessage\\"
    }
  },
  styles: [{
    name: \\"default\\",
    label: \\"Default\\",
    isDefault: true
  }, {
    name: \\"other\\",
    label: \\"Other\\"
  }]
};
"
`;
